Signal/src/ViewControllers/MessageMetadataViewController.swift

```diff
@@ -0,0 +1,393 @@
+                                 value:""))
+            return rows
+        }
+
+        let fileSize = dataSource.dataLength()
+        rows.append(valueRow(name: NSLocalizedString("MESSAGE_METADATA_VIEW_ATTACHMENT_FILE_SIZE",
+                                                     comment: "Label for file size of attachments in the 'message metadata' view."),
+                             value:ViewControllerUtils.formatFileSize(UInt(fileSize))))
+
+        if let dataUTI = MIMETypeUtil.utiType(forMIMEType:contentType) {
+            if attachment.isVoiceMessage() {
+                rows.append(valueRow(name: NSLocalizedString("MESSAGE_METADATA_VIEW_VOICE_MESSAGE",
+                                                             comment: "Label for voice messages of the 'message metadata' view."),
+                                     value:""))
+            } else {
+                rows.append(valueRow(name: NSLocalizedString("MESSAGE_METADATA_VIEW_MEDIA",
+                                                             comment: "Label for media messages of the 'message metadata' view."),
+                                     value:""))
+            }
+            let attachment = SignalAttachment(dataSource : dataSource, dataUTI: dataUTI)
+            let mediaMessageView = MediaMessageView(attachment:attachment)
+            self.mediaMessageView = mediaMessageView
+            rows.append(mediaMessageView)
+        }
+        return rows
+    }
+
+    private func recipientStatus(forOutgoingMessage message: TSOutgoingMessage, recipientId: String) -> String {
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateStyle = .short
+        dateFormatter.timeStyle = .long
+
+        let recipientReadMap = message.recipientReadMap
+        if let readTimestamp = recipientReadMap[recipientId] {
+            assert(message.messageState == .sentToService)
+            let readDate = NSDate.ows_date(withMillisecondsSince1970:readTimestamp.uint64Value)
+            return String(format:NSLocalizedString("MESSAGE_STATUS_READ_WITH_TIMESTAMP_FORMAT",
+                                                   comment: "message status for messages read by the recipient. Embeds: {{the date and time the message was read}}."),
+                          dateFormatter.string(from:readDate))
+        }
+
+        // TODO: We don't currently track delivery state on a per-recipient basis.
+        //       We should.
+        if message.wasDelivered {
+            return NSLocalizedString("MESSAGE_STATUS_DELIVERED",
+                                     comment:"message status for message delivered to their recipient.")
+        }
+
+        if message.messageState == .unsent {
+            return NSLocalizedString("MESSAGE_STATUS_FAILED", comment:"message footer for failed messages")
+        } else if (message.messageState == .sentToService ||
```

Comment:
nit: can you add a comment for the ages

```
// For legacy messages, we'll show as sent to all recipients even though it might have only been sent to one.
```

Reply:
I've added a comment.